### PR TITLE
Implement max31875 sensor

### PR DESCRIPTION
**Pull request by pdietl**

### CODE REVIEW

File: drivers/sensor/max31875/Kconfig [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# MAX31875 temperature sensor configuration options
+
+# Copyright (c) 2021 Firmwave
```
Comment (pdietl) [inline drivers/sensor/max31875/Kconfig:3]: Should I put my company here or...?

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -592,7 +592,14 @@ static int start_read(const struct device *dev,
...
 	!defined(STM32F3X_ADC_V2_5) && \
 	!defined(CONFIG_SOC_SERIES_STM32L1X)
+
+		/* we cannot calibrate the ADC while the ADC is enabled */
+		LL_ADC_Disable(adc);
+		while (LL_ADC_IsEnabled(adc) == 1UL) {
+		}
 		adc_stm32_calib(dev);
+		/* re-enable ADC after calibration */
+		LL_ADC_Enable(adc);
```
Comment (pdietl) [inline drivers/adc/adc_stm32.c:602]: These changes are part of another PR https://github.com/zephyrproject-rtos/zephyr/pull/40937 that just needs to be merged

File: drivers/sensor/max31875/max31875.c [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+ *
+ * @param mode  Data format
+ * @param res   Resolution (number of bits)
+ * @param convs Conversions per second
+ */
+#define MAX31875_CONFIG(format, \
+			res, \
+			convs) \
+	(((format) << (MAX31875_DATA_FORMAT_SHIFT)) | ((res) << (MAX31875_RESOLUTION_SHIFT)) | \
+	((convs) << (MAX31875_CONV_PER_SEC_SHIFT)))
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:60]: please format this better,
```c
#define MAX31875_CONFIG(format, res, convs) \
	(((format) << (MAX31875_DATA_FORMAT_SHIFT)) | ((res) << (MAX31875_RESOLUTION_SHIFT)) | \
	((convs) << (MAX31875_CONV_PER_SEC_SHIFT)))
```

File: drivers/sensor/max31875/max31875.c [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+struct max31875_data {
+	int16_t sample;
+	uint16_t config_reg;
+};
+
+struct max31875_config {
+	const struct i2c_dt_spec bus;
+	uint8_t conversions_per_second : 2,
+		data_format : 1,
+		resolution : 2;
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:74]: better use `uint8_t` for each entry

File: drivers/sensor/max31875/max31875.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+};
+
+static int max31875_reg_read(const struct max31875_config *cfg,
+			   uint8_t reg, uint16_t *val)
+{
+	if (i2c_burst_read_dt(&cfg->bus, reg, (uint8_t *)val, sizeof(*val)) < 0) {
+		return -EIO;
+	}
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:82]: better propagate error codes.

File: drivers/sensor/max31875/max31875.c [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+			value = MAX31875_DATA_FORMAT_NORMAL << MAX31875_DATA_FORMAT_SHIFT;
+		} else if (val->val1 == 150) {
+			value = MAX31875_DATA_FORMAT_EXTENDED << MAX31875_DATA_FORMAT_SHIFT;
+		} else {
+			return -ENOTSUP;
+		}
+
+		if (max31875_update_config(dev, MAX31875_DATA_FORMAT_SHIFT, value) < 0) {
+			LOG_DBG("Failed to set attribute!");
+			return -EIO;
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:145]: `LOG_ERR`, error code propagation

File: drivers/sensor/max31875/max31875.c [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+			break;
+
+		default:
+			return -ENOTSUP;
+		}
+
+		if (max31875_update_config(dev, MAX31875_CONV_PER_SEC_MASK, value) < 0) {
+			LOG_DBG("Failed to set attribute!");
+			return -EIO;
+		}
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:178]: ditto

File: drivers/sensor/max31875/max31875.c [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+static int max31875_sample_fetch(const struct device *dev,
+			       enum sensor_channel chan)
+{
+	struct max31875_data *data = dev->data;
+	const struct max31875_config *cfg = dev->config;
+	uint16_t val;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_AMBIENT_TEMP);
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:196]: better return -ENOTSUP if not true.

File: drivers/sensor/max31875/max31875.c [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+	if (max31875_reg_read(cfg, MAX31875_REG_TEMPERATURE, &val) < 0) {
+		return -EIO;
+	}
+
+	if (data->config_reg & (1 << MAX31875_DATA_FORMAT_SHIFT)) {
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:202]: `BIT()`

File: drivers/sensor/max31875/max31875.c [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+};
+
+int max31875_init(const struct device *dev)
+{
+	const struct max31875_config *cfg = dev->config;
+	struct max31875_data *data = dev->data;
+
+	if (!device_is_ready(cfg->bus.bus)) {
+		LOG_ERR("I2C dev %s not ready", cfg->bus.bus->name);
+		return -EINVAL;
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:244]: `-ENODEV`

File: drivers/sensor/max31875/max31875.c [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+	return max31875_update_config(dev, 0, 0);
+}
+
+
+#define MAX31875_INST(inst) \
+	static struct max31875_data max31875_data_##inst; \
+	static const struct max31875_config max31875_config_##inst = { \
+		.bus = I2C_DT_SPEC_INST_GET(inst), \
+		.conversions_per_second = DT_ENUM_IDX(DT_DRV_INST(inst), conversions_per_second), \
+		.resolution = DT_ENUM_IDX(DT_DRV_INST(inst), resolution), \
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:259]: `DT_INST_ENUM_IDX`

File: dts/bindings/sensor/maxim,max31875.yaml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+      - "8"
+  extended-mode:
+    description: When true, use 13-bit data format to allow for measuring temperatures > 128°C
+    type: boolean
+  resolution:
+    description: |
+      The number of bits used for each temperature sample. A resolution of 10 bits takes 35ms to convert.
+      Each 1-bit increase doubles the conversion time. Similarly, each 1-bit decrease halves the conversion time.
+    type: int
+    default: 10
```
Comment (gmarull) [inline dts/bindings/sensor/maxim,max31875.yaml:27]: defaults need to be justified. is sensor default? please update descriptions informing about that.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/sensor/max31875/Kconfig (hunk 1 above)
Comment (gmarull) [reply drivers/sensor/max31875/Kconfig:3]: who owns the copyright of this file :-)

File: drivers/sensor/max31875/max31875.c (hunk 11 above)
Comment (pdietl) [reply drivers/sensor/max31875/max31875.c:259]: good one!

Comment (pdietl) [reply drivers/sensor/max31875/max31875.c:259]: actually the build fails with `DT_INST_ENUM_IDX`

Comment (pdietl) [reply drivers/sensor/max31875/max31875.c:259]: or I just need to replace with `DT_INST_ENUM_IDX(inst, resolution)`

File: dts/bindings/sensor/maxim,max31875.yaml [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  Maxim MAX31875 Low-Power I2C Temperature Sensor.
+  Find the datasheet here https://datasheets.maximintegrated.com/en/ds/MAX31875.pdf
+
+compatible: "maxim,max31875"
+
+include: "i2c-device.yaml"
+
+properties:
+  conversions-per-second:
+    description: Number of temperature readings performed by the MAX31875 per second.
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/maxim,max31875.yaml:11]: All use of defaults must be justified in the description. Please read https://docs.zephyrproject.org/latest/guides/dts/bindings.html#rules-for-mainline-bindings in detail for all the rules about writing bindings for submission upstream.

File: dts/bindings/sensor/maxim,max31875.yaml (hunk 12 above)
Comment (mbolivar-nordic) [reply dts/bindings/sensor/maxim,max31875.yaml:27]: > This is not the sensor default. I chose it arbitrarily.

That's not OK -- see the details in the link I pasted earlier for the rules on default. We only allow it in limited situations.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
There are some issues with the binding.

**Comment by pdietl** [comment]
@mbolivar-nordic Changes defaults to power-on reset configuration

**Comment by erwango** [comment]
@pdietl You can rebase this PR, now that ADC change was merged

**Comment by pdietl** [comment]
> @pdietl You can rebase this PR, now that ADC change was merged

Should be dene @erwango

File: drivers/sensor/max31875/max31875.c [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+
+struct max31875_config {
+	const struct i2c_dt_spec bus;
+	uint8_t conversions_per_second;
+	uint8_t	data_format;
+	uint8_t	resolution;
+};
+
+static int max31875_reg_read(const struct max31875_config *cfg,
+			   uint8_t reg, uint16_t *val)
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:76]:
```suggestion
static int max31875_reg_read(const struct max31875_config *cfg,
			     uint8_t reg, uint16_t *val)
```

File: drivers/sensor/max31875/max31875.c [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+{
+	int ret = i2c_burst_read_dt(&cfg->bus, reg, (uint8_t *)val, sizeof(*val));
+
+	if (ret < 0) {
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:80]:
```suggestion
	int ret = i2c_burst_read_dt(&cfg->bus, reg, (uint8_t *)val, sizeof(*val));
	if (ret < 0) {
```

File: drivers/sensor/max31875/max31875.c [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+		return ret;
+	}
+
+	*val = sys_be16_to_cpu(*val);
+	return 0;
+}
+
+static int max31875_reg_write(const struct max31875_config *cfg,
+		uint8_t reg, uint16_t val)
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:89]:
```suggestion
static int max31875_reg_write(const struct max31875_config *cfg,
			      uint8_t reg, uint16_t val)
```

File: drivers/sensor/max31875/max31875.c [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+{
+	uint16_t val_be = sys_cpu_to_be16(val);
+
+	return i2c_burst_write_dt(&cfg->bus, reg, (uint8_t *)&val_be, 2);
+}
+
+static uint16_t set_config_flags(struct max31875_data *data, uint16_t mask,
+		uint16_t value)
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:97]:
```suggestion
static uint16_t set_config_flags(struct max31875_data *data, uint16_t mask,
				 uint16_t value)
```

File: drivers/sensor/max31875/max31875.c [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+		data->config_reg = new_val;
+	}
+
+	return rc;
+}
+
+static int max31875_attr_set(const struct device *dev,
+			   enum sensor_channel chan,
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:122]:
```suggestion
static int max31875_attr_set(const struct device *dev,
			     enum sensor_channel chan,
			     enum sensor_attribute attr,
			     const struct sensor_value *val)
```

File: drivers/sensor/max31875/max31875.c [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+	uint16_t val;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_AMBIENT_TEMP) {
+		LOG_ERR("Invalid channel provided");
+		return -ENOTSUP;
+	}
+
+	if (max31875_reg_read(cfg, MAX31875_REG_TEMPERATURE, &val) < 0) {
+		return -EIO;
+	}
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:206]:
```suggestion
	ret = max31875_reg_read(cfg, MAX31875_REG_TEMPERATURE, &val);
	if (ret < 0) {
		return ret;
	}
```

File: drivers/sensor/max31875/max31875.c [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+
+	if (data->config_reg & BIT(MAX31875_DATA_FORMAT_SHIFT)) {
+		data->sample = arithmetic_shift_right((int16_t)val,
+							  MAX31875_DATA_FORMAT_EXTENDED_SHIFT);
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:210]:
```suggestion
		data->sample = arithmetic_shift_right((int16_t)val,
						        MAX31875_DATA_FORMAT_EXTENDED_SHIFT);
```

File: drivers/sensor/max31875/max31875.c [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+	} else {
+		data->sample = arithmetic_shift_right((int16_t)val,
+							  MAX31875_DATA_FORMAT_NORMAL_SHIFT);
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:213]:
```suggestion
		data->sample = arithmetic_shift_right((int16_t)val,
						       MAX31875_DATA_FORMAT_NORMAL_SHIFT);
```

File: drivers/sensor/max31875/max31875.c [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+	const struct max31875_config *cfg = dev->config;
+	struct max31875_data *data = dev->data;
+
+	if (!device_is_ready(cfg->bus.bus)) {
+		LOG_ERR("I2C dev %s not ready", cfg->bus.bus->name);
+		return -ENODEV;
+	}
+
+	data->config_reg = MAX31875_CONFIG(cfg->data_format, cfg->resolution,
+		cfg->conversions_per_second);
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:254]:
```suggestion
	data->config_reg = MAX31875_CONFIG(cfg->data_format, cfg->resolution,
					   cfg->conversions_per_second);
```

File: drivers/sensor/max31875/max31875.c [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+
+	return max31875_update_config(dev, 0, 0);
+}
+
+
+#define MAX31875_INST(inst) \
+	static struct max31875_data max31875_data_##inst; \
+	static const struct max31875_config max31875_config_##inst = { \
+		.bus = I2C_DT_SPEC_INST_GET(inst), \
+		.conversions_per_second = DT_ENUM_IDX(DT_DRV_INST(inst), conversions_per_second), \
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:264]:
```suggestion
		.conversions_per_second = DT_INST_ENUM_IDX(inst, conversions_per_second), \
```

File: drivers/sensor/max31875/max31875.c [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+		.resolution = DT_INST_ENUM_IDX(inst, resolution), \
+		.data_format = DT_INST_PROP(inst, extended_mode), \
+	};\
+	DEVICE_DT_INST_DEFINE(inst, max31875_init, NULL, &max31875_data_##inst, \
+				&max31875_config_##inst, POST_KERNEL, \
+				CONFIG_SENSOR_INIT_PRIORITY, &max31875_driver_api);
```
Comment (gmarull) [inline drivers/sensor/max31875/max31875.c:270]: by convention: push all `\` to 80 cols (or align them to the same column at least)

File: dts/bindings/sensor/maxim,max31875.yaml [hunk 25]
```diff
@@ -0,0 +1,35 @@
+description:
```
Comment (gmarull) [inline dts/bindings/sensor/maxim,max31875.yaml:1]:
```suggestion
description: |
```

File: dts/bindings/sensor/maxim,max31875.yaml [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+      - "1"
+      - "4"
+      - "8"
+  extended-mode:
+    description: When true, use 13-bit data format to allow for measuring temperatures > 128°C
+    type: boolean
+  resolution:
+    description: |
+      The number of bits used for each temperature sample. A resolution of 10 bits takes 35ms to convert.
+      Each 1-bit increase doubles the conversion time. Similarly, each 1-bit decrease halves the conversion time.
```
Comment (gmarull) [inline dts/bindings/sensor/maxim,max31875.yaml:27]: default not justified here

**Review by gmarull — CHANGES_REQUESTED** [review]

**Comment by pdietl** [comment]
@gmarull I ran `uncrustify` :p

**Comment by pdietl** [comment]
@gmarull @mbolivar-nordic I think I addressed everything :)

**Comment by pdietl** [comment]
poke @gmarull @mbolivar-nordic

**Review by gmarull — COMMENTED** [review]
please add this sensor to tests/drivers/build_all/sensor/ in a new commit

**Review by tejlmand — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]
Happy new year!

Can you please line wrap the long description fields in the binding to something smaller, say 80 columns?

This is not a blocking request, but with these very long lines, you will be forcing readers to do a fair bit of horizontal scrolling in the bindings index (https://docs.zephyrproject.org/latest/reference/devicetree/bindings.html) page for your binding.

People hate horizontal scrolling.

**Comment by pdietl** [comment]
@gmarull Is this what you'd like? 
```
diff --git a/tests/drivers/build_all/sensor/prj.conf b/tests/drivers/build_all/sensor/prj.conf
index f2ac144eb7..120db1876a 100644
--- a/tests/drivers/build_all/sensor/prj.conf
+++ b/tests/drivers/build_all/sensor/prj.conf
@@ -104,3 +104,4 @@ CONFIG_INA23X=y
 CONFIG_INA230=y
 CONFIG_INA237=y
 CONFIG_MHZ19B=y
+CONFIG_MAX31875=y
```

?

I tried building the tests with `./scripts/twister -s tests/drivers/build_all/sensor/sensors.build` but it failed?

**Comment by pdietl** [comment]
@mbolivar-nordic  I wrapped the lines

**Comment by pdietl** [comment]
poke @gmarull

**Review by gmarull — CHANGES_REQUESTED** [review]
Please add the driver to the sensors build_all test

File: dts/bindings/sensor/maxim,max31875.yaml [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
```
Comment (gmarull) [inline dts/bindings/sensor/maxim,max31875.yaml:27]: wrap to 80cols?

**Comment by pdietl** [comment]
@gmarull I commented earlier a question to you. I will repeat it here:



> Is this what you'd like?
> 
> ```
> diff --git a/tests/drivers/build_all/sensor/prj.conf b/tests/drivers/build_all/sensor/prj.conf
> index f2ac144eb7..120db1876a 100644
> --- a/tests/drivers/build_all/sensor/prj.conf
> +++ b/tests/drivers/build_all/sensor/prj.conf
> @@ -104,3 +104,4 @@ CONFIG_INA23X=y
>  CONFIG_INA230=y
>  CONFIG_INA237=y
>  CONFIG_MHZ19B=y
> +CONFIG_MAX31875=y
> ```
> 
> ?
> 
> I tried building the tests with `./scripts/twister -s tests/drivers/build_all/sensor/sensors.build` but it failed?

File: drivers/sensor/max31875/max31875.c [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+	return 0;
+}
+
+static const struct sensor_driver_api max31875_driver_api = {
+	.attr_set = max31875_attr_set,
+	.sample_fetch = max31875_sample_fetch,
+	.channel_get = max31875_channel_get,
+};
+
+int max31875_init(const struct device *dev)
```
Comment (MaureenHelm) [inline drivers/sensor/max31875/max31875.c:246]: Make this function `static`

File: dts/bindings/sensor/maxim,max31875.yaml [hunk 29]
```diff
@@ -0,0 +1,36 @@
+description: |
```
Comment (MaureenHelm) [inline dts/bindings/sensor/maxim,max31875.yaml:1]: Add license/copyright header

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> @gmarull I commented earlier a question to you. I will repeat it here:
> 
> Is this what you'd like?
> 
> ```diff
> diff --git a/tests/drivers/build_all/sensor/prj.conf b/tests/drivers/build_all/sensor/prj.conf
> index f2ac144eb7..120db1876a 100644
> --- a/tests/drivers/build_all/sensor/prj.conf
> +++ b/tests/drivers/build_all/sensor/prj.conf
> @@ -104,3 +104,4 @@ CONFIG_INA23X=y
>  CONFIG_INA230=y
>  CONFIG_INA237=y
>  CONFIG_MHZ19B=y
> +CONFIG_MAX31875=y
>```
> ?
>
> I tried building the tests with ./scripts/twister -s tests/drivers/build_all/sensor/sensors.build but it failed?

Yes, this is what we want because it ensures that the new driver builds in CI. How did the test fail?

**Comment by pdietl** [comment]
I need some help debugging the Twister issues @gmarull.

**Comment by gmarull** [comment]
Have you updated the corresponding DTS file in tests? https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/sensor/i2c.dtsi

**Review by gmarull — DISMISSED** [review]

**Comment by pdietl** [comment]
poke @MaureenHelm I think I addressed your comments.

**Comment by pdietl** [comment]
@galak  @nashif @MaureenHelm are we good to merge?

File: dts/bindings/sensor/maxim,max31875.yaml (hunk 29 above)
Comment (MaureenHelm) [reply dts/bindings/sensor/maxim,max31875.yaml:7]: petedietl@gmail.h?

**Review by MaureenHelm — COMMENTED** [review]
There's a typo in your license/copyright header, but otherwise this looks good. We have to wait for the merge window to reopen after the 3.0 release.

See https://github.com/zephyrproject-rtos/zephyr/wiki/Program-Management for the schedule.

**Comment by pdietl** [comment]
@MaureenHelm Fixed the typo (I think I was writing too much C code). Thanks for the info. Now we wait.

**Review by MaureenHelm — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]
For the binding

**Comment by pdietl** [comment]
What do you mean?

**Comment by mbolivar-nordic** [comment]
> What do you mean?

My approval is for the binding only

**Comment by pdietl** [comment]
@MaureenHelm when is the merge window going to open?

**Comment by mbolivar-nordic** [comment]
> @MaureenHelm when is the merge window going to open?

https://github.com/zephyrproject-rtos/zephyr/pull/41107#pullrequestreview-866787334

**Comment by pdietl** [comment]
@mbolivar-nordic There is no M0 or M1 date for 3.1 listed though?

**Comment by mbolivar-nordic** [comment]
> @mbolivar-nordic There is no M0 or M1 date for 3.1 listed though?

True! Strange...

**Review by tejlmand — APPROVED** [review]
lgtm - reapprove

**Review by mbolivar-nordic — APPROVED** [review]
@MaureenHelm if you're happy with this, please re-approve and merge

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**